sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1242,6 +1243,53 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
 }
 
 /**
+ * If one or both of the datasets in the logical [[Except]] operator are purely transformed using
+ * [[Filter]], this rule will replace logical [[Except]] operator with a [[Filter]] operator by
+ * flipping the filter condition of the right child.
+ * {{{
+ *   SELECT a1, a2 FROM Tab1 WHERE a2 = 12 EXCEPT SELECT a1, a2 FROM Tab1 WHERE a1 = 5
+ *   ==>  SELECT a1, a2 FROM Tab1 WHERE a2 = 12 AND a1 <> 5
```

Comment:
This is wrong. See the correct description of `EXCEPT`

> EXCEPT returns distinct rows from the left input query that aren’t output by the right input query.

Reply:
@gatorsmile Thanks for the comment. Sorry i didn't pay attention, i'll update the query example. But the code produces the correct behaviour of the Except operator. Do you have any comment in the code?

Reply:
I can see the value of this rule. 

cc @gengliangwang Review this?